<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Augustus: GeneMSA Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Augustus
   &#160;<span id="projectnumber">3.4.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classGeneMSA-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">GeneMSA Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>multiple sequence alignment of genomes for comparative gene prediction  
 <a href="classGeneMSA.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="geneMSA_8hh_source.html">geneMSA.hh</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ada79b45bf3f2605773220304968d5993"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneMSA.html#ada79b45bf3f2605773220304968d5993">GeneMSA</a> (<a class="el" href="classRandSeqAccess.html">RandSeqAccess</a> *rsa, <a class="el" href="classAlignment.html">Alignment</a> *a)</td></tr>
<tr class="separator:ada79b45bf3f2605773220304968d5993"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a098d255c766990fe17b2cab50ddd99b9"><td class="memItemLeft" align="right" valign="top"><a id="a098d255c766990fe17b2cab50ddd99b9"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>getSeqID</b> (int speciesIdx)</td></tr>
<tr class="separator:a098d255c766990fe17b2cab50ddd99b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4f6721c32df049be3093fc3e61bfee7f"><td class="memItemLeft" align="right" valign="top"><a id="a4f6721c32df049be3093fc3e61bfee7f"></a>
Strand&#160;</td><td class="memItemRight" valign="bottom"><b>getStrand</b> (int speciesIdx)</td></tr>
<tr class="separator:a4f6721c32df049be3093fc3e61bfee7f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3553dfabdf1eb22999f02ebcd4bc150a"><td class="memItemLeft" align="right" valign="top"><a id="a3553dfabdf1eb22999f02ebcd4bc150a"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>getStart</b> (int speciesIdx)</td></tr>
<tr class="separator:a3553dfabdf1eb22999f02ebcd4bc150a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5851fd32db6fd14116bfe70d7f651d6d"><td class="memItemLeft" align="right" valign="top"><a id="a5851fd32db6fd14116bfe70d7f651d6d"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>getEnd</b> (int speciesIdx)</td></tr>
<tr class="separator:a5851fd32db6fd14116bfe70d7f651d6d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c1a22fb5aa0a7d1978406b354f43b62"><td class="memItemLeft" align="right" valign="top"><a id="a7c1a22fb5aa0a7d1978406b354f43b62"></a>
list&lt; <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; *&#160;</td><td class="memItemRight" valign="bottom"><b>getExonCands</b> (int speciesIdx)</td></tr>
<tr class="separator:a7c1a22fb5aa0a7d1978406b354f43b62"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3af1af84efc30d91569ff2fc07f877dc"><td class="memItemLeft" align="right" valign="top"><a id="a3af1af84efc30d91569ff2fc07f877dc"></a>
<a class="el" href="classAlignment.html">Alignment</a> *&#160;</td><td class="memItemRight" valign="bottom"><b>getAlignment</b> ()</td></tr>
<tr class="separator:a3af1af84efc30d91569ff2fc07f877dc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1cbb56d4306431c2ff9c99ea4e9c0ca2"><td class="memItemLeft" align="right" valign="top"><a id="a1cbb56d4306431c2ff9c99ea4e9c0ca2"></a>
vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>getOffsets</b> ()</td></tr>
<tr class="separator:a1cbb56d4306431c2ff9c99ea4e9c0ca2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d2bab02c1055fcf87569b4e81dbf473"><td class="memItemLeft" align="right" valign="top"><a id="a2d2bab02c1055fcf87569b4e81dbf473"></a>
map&lt; string, <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; *&#160;</td><td class="memItemRight" valign="bottom"><b>getECHash</b> (list&lt; <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; *ec)</td></tr>
<tr class="separator:a2d2bab02c1055fcf87569b4e81dbf473"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06a64530efb0f34260955e2c2f15b6e1"><td class="memItemLeft" align="right" valign="top"><a id="a06a64530efb0f34260955e2c2f15b6e1"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>createExonCands</b> (int s, const char *dna, map&lt; int_fast64_t, <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; &amp;ecs, map&lt; int_fast64_t, <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; &amp;addECs)</td></tr>
<tr class="separator:a06a64530efb0f34260955e2c2f15b6e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc65c9a4e28f980100508ad78948a80d"><td class="memItemLeft" align="right" valign="top"><a id="abc65c9a4e28f980100508ad78948a80d"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>setExonCands</b> (vector&lt; map&lt; int_fast64_t, <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; &gt; &amp;ecs)</td></tr>
<tr class="separator:abc65c9a4e28f980100508ad78948a80d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad55bc5eb24509be6b5aa056e8db5ef1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneMSA.html#aad55bc5eb24509be6b5aa056e8db5ef1">createOrthoExons</a> (list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; &amp;orthoExonsList, map&lt; int_fast64_t, list&lt; pair&lt; int, <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; &gt; &gt; &amp;alignedECs, <a class="el" href="classEvo.html">Evo</a> *evo, float consThres=0.0, int minAvLen=0)</td></tr>
<tr class="separator:aad55bc5eb24509be6b5aa056e8db5ef1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1978412722f890d5391a9e4b109a059b"><td class="memItemLeft" align="right" valign="top"><a id="a1978412722f890d5391a9e4b109a059b"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>printStats</b> ()</td></tr>
<tr class="separator:a1978412722f890d5391a9e4b109a059b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad7fe754bf4dba15062935d9384dc7297"><td class="memItemLeft" align="right" valign="top"><a id="ad7fe754bf4dba15062935d9384dc7297"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>printGeneRanges</b> ()</td></tr>
<tr class="separator:ad7fe754bf4dba15062935d9384dc7297"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4879002498c63d42779e3602d6a5957e"><td class="memItemLeft" align="right" valign="top"><a id="a4879002498c63d42779e3602d6a5957e"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>printExonCands</b> ()</td></tr>
<tr class="separator:a4879002498c63d42779e3602d6a5957e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af4640c957ad2881ae470e3233cbbeebe"><td class="memItemLeft" align="right" valign="top"><a id="af4640c957ad2881ae470e3233cbbeebe"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>printOrthoExons</b> (list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; &amp;orthoExonsList)</td></tr>
<tr class="separator:af4640c957ad2881ae470e3233cbbeebe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0116ba0a45565b8eeabc7723ceca0191"><td class="memItemLeft" align="right" valign="top"><a id="a0116ba0a45565b8eeabc7723ceca0191"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>computeOmegas</b> (list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; &amp;orthoExonsList, vector&lt; <a class="el" href="classAnnoSequence.html">AnnoSequence</a> * &gt; const &amp;seqRanges, <a class="el" href="classPhyloTree.html">PhyloTree</a> *ctree)</td></tr>
<tr class="separator:a0116ba0a45565b8eeabc7723ceca0191"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a95562d24d0e41aa502810e6e386ed634"><td class="memItemLeft" align="right" valign="top"><a id="a95562d24d0e41aa502810e6e386ed634"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>computeOmegasEff</b> (list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; &amp;orthoExonsList, vector&lt; <a class="el" href="classAnnoSequence.html">AnnoSequence</a> * &gt; const &amp;seqRanges, <a class="el" href="classPhyloTree.html">PhyloTree</a> *ctree, ofstream *codonAli)</td></tr>
<tr class="separator:a95562d24d0e41aa502810e6e386ed634"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acd216a18af2ed9013758eb933874fefa"><td class="memItemLeft" align="right" valign="top"><a id="acd216a18af2ed9013758eb933874fefa"></a>
vector&lt; string &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>pruneToBV</b> (vector&lt; string &gt; *cs, bit_vector bv)</td></tr>
<tr class="separator:acd216a18af2ed9013758eb933874fefa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac2ee12213e0164c4bf34dca6b293a3c1"><td class="memItemLeft" align="right" valign="top"><a id="ac2ee12213e0164c4bf34dca6b293a3c1"></a>
vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>pruneToBV</b> (vector&lt; int &gt; *rfc, bit_vector bv)</td></tr>
<tr class="separator:ac2ee12213e0164c4bf34dca6b293a3c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a71f77b5f688d7d565be95a0df95d765e"><td class="memItemLeft" align="right" valign="top"><a id="a71f77b5f688d7d565be95a0df95d765e"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>omegaForCodonTuple</b> (vector&lt; double &gt; *loglik)</td></tr>
<tr class="separator:a71f77b5f688d7d565be95a0df95d765e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6240bdfda3749c770170c7be830d2ac8"><td class="memItemLeft" align="right" valign="top"><a id="a6240bdfda3749c770170c7be830d2ac8"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>printOmegaForCodon</b> (string outdir)</td></tr>
<tr class="separator:a6240bdfda3749c770170c7be830d2ac8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a7d243458c52f3962550cf8b070a8d4"><td class="memItemLeft" align="right" valign="top"><a id="a0a7d243458c52f3962550cf8b070a8d4"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>printCumOmega</b> ()</td></tr>
<tr class="separator:a0a7d243458c52f3962550cf8b070a8d4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d952b92c2f0d05be21e47c4af2ff6d5"><td class="memItemLeft" align="right" valign="top"><a id="a1d952b92c2f0d05be21e47c4af2ff6d5"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>comparativeSignalScoring</b> (list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; &amp;orthoExonsList)</td></tr>
<tr class="separator:a1d952b92c2f0d05be21e47c4af2ff6d5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa852177a37532f35d6442cae1a9f84c7"><td class="memItemLeft" align="right" valign="top"><a id="aa852177a37532f35d6442cae1a9f84c7"></a>
<a class="el" href="classLocusTree.html">LocusTree</a> *&#160;</td><td class="memItemRight" valign="bottom"><b>constructTree</b> ()</td></tr>
<tr class="separator:aa852177a37532f35d6442cae1a9f84c7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5b637955eaf250492c9a4a88254cd38"><td class="memItemLeft" align="right" valign="top"><a id="af5b637955eaf250492c9a4a88254cd38"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>calcConsScore</b> (list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; &amp;orthoExonsList, vector&lt; <a class="el" href="classAnnoSequence.html">AnnoSequence</a> * &gt; const &amp;seqRanges, string outdir)</td></tr>
<tr class="separator:af5b637955eaf250492c9a4a88254cd38"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac23fa08db7e09bec955cf03024cc9a30"><td class="memItemLeft" align="right" valign="top"><a id="ac23fa08db7e09bec955cf03024cc9a30"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>calcColumnScore</b> (int a, int c, int t, int g)</td></tr>
<tr class="separator:ac23fa08db7e09bec955cf03024cc9a30"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97f77a909c854ae7354cf10c1dd23288"><td class="memItemLeft" align="right" valign="top"><a id="a97f77a909c854ae7354cf10c1dd23288"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>consToWig</b> (vector&lt; double &gt; &amp;consScore, string outdir)</td></tr>
<tr class="separator:a97f77a909c854ae7354cf10c1dd23288"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a22f8dc2bfde374dbdf0948d2da38532c"><td class="memItemLeft" align="right" valign="top"><a id="a22f8dc2bfde374dbdf0948d2da38532c"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>printSingleOrthoExon</b> (<a class="el" href="classOrthoExon.html">OrthoExon</a> &amp;oe, bool files=true)</td></tr>
<tr class="separator:a22f8dc2bfde374dbdf0948d2da38532c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2bcd499f2add9948ce335746f1610efa"><td class="memItemLeft" align="right" valign="top"><a id="a2bcd499f2add9948ce335746f1610efa"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>collect_features</b> (int species, list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; *hects, <a class="el" href="classSpeciesGraph.html">SpeciesGraph</a> *speciesgraph)</td></tr>
<tr class="separator:a2bcd499f2add9948ce335746f1610efa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae023b0e02bb2a77a2b87db9126384f91"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneMSA.html#ae023b0e02bb2a77a2b87db9126384f91">getAllOEMsas</a> (int species, list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; *hects, unordered_map&lt; string, int &gt; *ref_class, vector&lt; <a class="el" href="classAnnoSequence.html">AnnoSequence</a> * &gt; const &amp;seqRanges)</td></tr>
<tr class="separator:ae023b0e02bb2a77a2b87db9126384f91"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd42e0ecb942aceb3d0b7d7f4c104dce"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classStringAlignment.html">StringAlignment</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneMSA.html#abd42e0ecb942aceb3d0b7d7f4c104dce">getMsa</a> (<a class="el" href="classOrthoExon.html">OrthoExon</a> const &amp;oe, vector&lt; <a class="el" href="classAnnoSequence.html">AnnoSequence</a> * &gt; const &amp;seqRanges, size_t flanking=0)</td></tr>
<tr class="separator:abd42e0ecb942aceb3d0b7d7f4c104dce"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a8c5d80ca4bcd8afb80734af19df6e8bc"><td class="memItemLeft" align="right" valign="top"><a id="a8c5d80ca4bcd8afb80734af19df6e8bc"></a>
static void&#160;</td><td class="memItemRight" valign="bottom"><b>setTree</b> (<a class="el" href="classPhyloTree.html">PhyloTree</a> *t)</td></tr>
<tr class="separator:a8c5d80ca4bcd8afb80734af19df6e8bc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5158d5fceb4b40da3f92f211a04c8ab"><td class="memItemLeft" align="right" valign="top"><a id="af5158d5fceb4b40da3f92f211a04c8ab"></a>
static void&#160;</td><td class="memItemRight" valign="bottom"><b>setCodonEvo</b> (<a class="el" href="classCodonEvo.html">CodonEvo</a> *c)</td></tr>
<tr class="separator:af5158d5fceb4b40da3f92f211a04c8ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae29c97023cfdf801b529c99652fa4b1d"><td class="memItemLeft" align="right" valign="top"><a id="ae29c97023cfdf801b529c99652fa4b1d"></a>
static int&#160;</td><td class="memItemRight" valign="bottom"><b>numSpecies</b> ()</td></tr>
<tr class="separator:ae29c97023cfdf801b529c99652fa4b1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae33b7dd846abc77d3f00e2dd3697fc01"><td class="memItemLeft" align="right" valign="top"><a id="ae33b7dd846abc77d3f00e2dd3697fc01"></a>
static void&#160;</td><td class="memItemRight" valign="bottom"><b>openOutputFiles</b> (string outdir)</td></tr>
<tr class="separator:ae33b7dd846abc77d3f00e2dd3697fc01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3e5f25a0d2debe8d3168691f13b540ca"><td class="memItemLeft" align="right" valign="top"><a id="a3e5f25a0d2debe8d3168691f13b540ca"></a>
static void&#160;</td><td class="memItemRight" valign="bottom"><b>closeOutputFiles</b> ()</td></tr>
<tr class="separator:a3e5f25a0d2debe8d3168691f13b540ca"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a6221a8c329c4df34114d68ef49a3ef04"><td class="memItemLeft" align="right" valign="top"><a id="a6221a8c329c4df34114d68ef49a3ef04"></a>
static int&#160;</td><td class="memItemRight" valign="bottom"><b>padding</b> = 1000</td></tr>
<tr class="separator:a6221a8c329c4df34114d68ef49a3ef04"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb07b965c77fff35115a956fb590e2af"><td class="memItemLeft" align="right" valign="top"><a id="adb07b965c77fff35115a956fb590e2af"></a>
static int&#160;</td><td class="memItemRight" valign="bottom"><b>orthoExonID</b> = 1</td></tr>
<tr class="separator:adb07b965c77fff35115a956fb590e2af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06055a284db5024ea2b6f33f8ad1154a"><td class="memItemLeft" align="right" valign="top"><a id="a06055a284db5024ea2b6f33f8ad1154a"></a>
static int&#160;</td><td class="memItemRight" valign="bottom"><b>geneRangeID</b> = 1</td></tr>
<tr class="separator:a06055a284db5024ea2b6f33f8ad1154a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a62906316b181cbc34cae6aa6e134acea"><td class="memItemLeft" align="right" valign="top"><a id="a62906316b181cbc34cae6aa6e134acea"></a>
static vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>exonCandID</b></td></tr>
<tr class="separator:a62906316b181cbc34cae6aa6e134acea"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a19e47993c31031eb842fff7341af303a"><td class="memItemLeft" align="right" valign="top"><a id="a19e47993c31031eb842fff7341af303a"></a>
static unordered_map&lt; bit_vector, <a class="el" href="classPhyloTree.html">PhyloTree</a> *, boost::hash&lt; bit_vector &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>topologies</b></td></tr>
<tr class="separator:a19e47993c31031eb842fff7341af303a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6005299f66b11e5df33001520a544f27"><td class="memItemLeft" align="right" valign="top"><a id="a6005299f66b11e5df33001520a544f27"></a>
static vector&lt; ofstream * &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>exonCands_outfiles</b></td></tr>
<tr class="separator:a6005299f66b11e5df33001520a544f27"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb7c85f45bfeb3d26570d50812de9d78"><td class="memItemLeft" align="right" valign="top"><a id="abb7c85f45bfeb3d26570d50812de9d78"></a>
static vector&lt; ofstream * &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>orthoExons_outfiles</b></td></tr>
<tr class="separator:abb7c85f45bfeb3d26570d50812de9d78"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae43347c37e00c6e96347200294808bf0"><td class="memItemLeft" align="right" valign="top"><a id="ae43347c37e00c6e96347200294808bf0"></a>
static vector&lt; ofstream * &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>geneRanges_outfiles_bed</b></td></tr>
<tr class="separator:ae43347c37e00c6e96347200294808bf0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acae456b81e35fafc4b62fe0e86edb865"><td class="memItemLeft" align="right" valign="top"><a id="acae456b81e35fafc4b62fe0e86edb865"></a>
static vector&lt; ofstream * &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>geneRanges_outfiles_gff</b></td></tr>
<tr class="separator:acae456b81e35fafc4b62fe0e86edb865"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4a21f61b0a050b7631c5f742a7d90dc"><td class="memItemLeft" align="right" valign="top"><a id="ac4a21f61b0a050b7631c5f742a7d90dc"></a>
static vector&lt; ofstream * &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>omega_outfiles</b></td></tr>
<tr class="separator:ac4a21f61b0a050b7631c5f742a7d90dc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f6b51027c36a61ccea85f3e57fdb404"><td class="memItemLeft" align="right" valign="top"><a id="a7f6b51027c36a61ccea85f3e57fdb404"></a>
static map&lt; vector&lt; string &gt;, pair&lt; vector&lt; double &gt;, int &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>computedCumValues</b></td></tr>
<tr class="separator:a7f6b51027c36a61ccea85f3e57fdb404"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>multiple sequence alignment of genomes for comparative gene prediction </p>
<dl class="section author"><dt>Author</dt><dd>Mario Stanke </dd>
<dd>
Alexander Gebauer </dd></dl>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="ada79b45bf3f2605773220304968d5993"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ada79b45bf3f2605773220304968d5993">&#9670;&nbsp;</a></span>GeneMSA()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">GeneMSA::GeneMSA </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classRandSeqAccess.html">RandSeqAccess</a> *&#160;</td>
          <td class="paramname"><em>rsa</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classAlignment.html">Alignment</a> *&#160;</td>
          <td class="paramname"><em>a</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>construct the gene ranges now: simple copy. TODO: extend region when apparently part of the alignment is missing human ***********&mdash;******************* mouse *******----&mdash;******************- chicken ***********-------------------&mdash; ^ | extend range here</p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="aad55bc5eb24509be6b5aa056e8db5ef1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aad55bc5eb24509be6b5aa056e8db5ef1">&#9670;&nbsp;</a></span>createOrthoExons()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneMSA::createOrthoExons </td>
          <td>(</td>
          <td class="paramtype">list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>orthoExonsList</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">map&lt; int_fast64_t, list&lt; pair&lt; int, <a class="el" href="classExonCandidate.html">ExonCandidate</a> * &gt; &gt; &gt; &amp;&#160;</td>
          <td class="paramname"><em>alignedECs</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classEvo.html">Evo</a> *&#160;</td>
          <td class="paramname"><em>evo</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>consThres</em> = <code>0.0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>minAvLen</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>find all ortholog exon candidates, that are present in at least max(2, consThresh * m) where m &lt;= numSpecies is the number of species that are present in the alignment Only report OrthoExons oe with at least 'minAvLen' as average length of the exon candidates in oe. ortholog exon candidates:</p><ul>
<li>both splice sites align exactly</li>
<li>the exon candidate types agrees (single, rsingle, internal0, ...)</li>
<li>the phases at both boundaries agree (i.e. exon candidate types and length modulo 3) EC coordinates are region-based, as they are used in the <a class="el" href="classOrthoGraph.html" title="orthologous graphs for comparative gene prediction">OrthoGraph</a></li>
</ul>
<p>createOrthoExons </p>

</div>
</div>
<a id="ae023b0e02bb2a77a2b87db9126384f91"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae023b0e02bb2a77a2b87db9126384f91">&#9670;&nbsp;</a></span>getAllOEMsas()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneMSA::getAllOEMsas </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">list&lt; <a class="el" href="classOrthoExon.html">OrthoExon</a> &gt; *&#160;</td>
          <td class="paramname"><em>hects</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unordered_map&lt; string, int &gt; *&#160;</td>
          <td class="paramname"><em>ref_class</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; <a class="el" href="classAnnoSequence.html">AnnoSequence</a> * &gt; const &amp;&#160;</td>
          <td class="paramname"><em>seqRanges</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>getAllOEMsas obtains and prints multiple sequence alignments (MSAs) and their label y=0,1, whether it constitutes a real CDS or not in the reference species.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">species</td><td>is the reference species index </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">hects</td><td>ortho exons to get alignments from </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_class</td><td>assignment of CDS keys as 1 (= in reference anno) or 0. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">seqRanges</td><td>the aligned sequences </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="abd42e0ecb942aceb3d0b7d7f4c104dce"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abd42e0ecb942aceb3d0b7d7f4c104dce">&#9670;&nbsp;</a></span>getMsa()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classStringAlignment.html">StringAlignment</a> GeneMSA::getMsa </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classOrthoExon.html">OrthoExon</a> const &amp;&#160;</td>
          <td class="paramname"><em>oe</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; <a class="el" href="classAnnoSequence.html">AnnoSequence</a> * &gt; const &amp;&#160;</td>
          <td class="paramname"><em>seqRanges</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>flanking</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>getMsa obtains the alignment in string format, including unaligned insertions that are not part of any fragment.</p>
<dl class="section return"><dt>Returns</dt><dd>msa vector of alignment rows </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">oe</td><td>the <a class="el" href="classOrthoExon.html" title="maintains orthologous exons for comparative gene prediction">OrthoExon</a> whose alignment is sought </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">seqRanges</td><td>contains the alignment of the larger region tuple </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">flanking</td><td>allows to add padding on both sides, e.g. for sequence signals </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>include/<a class="el" href="geneMSA_8hh_source.html">geneMSA.hh</a></li>
<li>src/geneMSA.cc</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.0
</small></address>
</body>
</html>
