<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Augustus: include/geneMSA.hh Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Augustus
   &#160;<span id="projectnumber">3.4.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_d44c64559bbebec7f509842c48db8b23.html">include</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">geneMSA.hh</div>  </div>
</div><!--header-->
<div class="contents">
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;<span class="comment">/*</span></div>
<div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;<span class="comment"> * geneMSA.hh</span></div>
<div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;<span class="comment"> *</span></div>
<div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;<span class="comment"> * License: Artistic License, see file LICENSE.TXT or </span></div>
<div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;<span class="comment"> *          https://opensource.org/licenses/artistic-license-1.0</span></div>
<div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;<span class="comment"> */</span></div>
<div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160; </div>
<div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;<span class="preprocessor">#ifndef _GENEMSA</span></div>
<div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;<span class="preprocessor">#define _GENEMSA</span></div>
<div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160; </div>
<div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;<span class="comment">// project includes</span></div>
<div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;<span class="preprocessor">#include &quot;alignment.hh&quot;</span></div>
<div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;<span class="preprocessor">#include &quot;exoncand.hh&quot;</span></div>
<div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;<span class="preprocessor">#include &quot;orthoexon.hh&quot;</span></div>
<div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;<span class="preprocessor">#include &quot;randseqaccess.hh&quot;</span></div>
<div class="line"><a name="l00016"></a><span class="lineno">   16</span>&#160;<span class="preprocessor">#include &quot;speciesgraph.hh&quot;</span></div>
<div class="line"><a name="l00017"></a><span class="lineno">   17</span>&#160;<span class="preprocessor">#include &quot;contTimeMC.hh&quot;</span></div>
<div class="line"><a name="l00018"></a><span class="lineno">   18</span>&#160;<span class="preprocessor">#include&lt;unordered_map&gt;</span></div>
<div class="line"><a name="l00019"></a><span class="lineno">   19</span>&#160;<span class="preprocessor">#include&lt;boost/functional/hash.hpp&gt;</span></div>
<div class="line"><a name="l00020"></a><span class="lineno">   20</span>&#160; </div>
<div class="line"><a name="l00021"></a><span class="lineno">   21</span>&#160;<span class="comment">//forward declarations</span></div>
<div class="line"><a name="l00022"></a><span class="lineno">   22</span>&#160; </div>
<div class="line"><a name="l00028"></a><span class="lineno">   28</span>&#160;<span class="keyword">class </span><a class="code" href="classOrthoGraph.html">OrthoGraph</a>;</div>
<div class="line"><a name="l00029"></a><span class="lineno">   29</span>&#160;<span class="keywordtype">string</span> printRFC(vector&lt;int&gt;);</div>
<div class="line"><a name="l00030"></a><span class="lineno">   30</span>&#160; </div>
<div class="line"><a name="l00031"></a><span class="lineno">   31</span>&#160;<span class="comment">/* store cumulative sums of log likelihoods for each omega (optional: cumulative sum of number of substitutions) </span></div>
<div class="line"><a name="l00032"></a><span class="lineno">   32</span>&#160;<span class="comment"> * we create an object of cumValues for each bit_vector and reading frame combination  </span></div>
<div class="line"><a name="l00033"></a><span class="lineno">   33</span>&#160;<span class="comment"> */</span></div>
<div class="line"><a name="l00034"></a><span class="lineno"><a class="line" href="structcumValues.html">   34</a></span>&#160;<span class="keyword">struct </span><a class="code" href="structcumValues.html">cumValues</a>{</div>
<div class="line"><a name="l00035"></a><span class="lineno">   35</span>&#160;  vector&lt;double&gt; logliks;</div>
<div class="line"><a name="l00036"></a><span class="lineno">   36</span>&#160;  <span class="keywordtype">int</span> numSubs;</div>
<div class="line"><a name="l00037"></a><span class="lineno">   37</span>&#160;  <span class="keywordtype">int</span> id;</div>
<div class="line"><a name="l00038"></a><span class="lineno">   38</span>&#160;  <a class="code" href="structcumValues.html">cumValues</a>(<span class="keywordtype">int</span> i, <span class="keywordtype">int</span> s=-1):numSubs(s), id(i){};</div>
<div class="line"><a name="l00039"></a><span class="lineno">   39</span>&#160;  <span class="keywordtype">void</span> addLogliks(vector&lt;double&gt;* ll){</div>
<div class="line"><a name="l00040"></a><span class="lineno">   40</span>&#160;    <span class="keywordflow">if</span>(logliks.size() == 0){</div>
<div class="line"><a name="l00041"></a><span class="lineno">   41</span>&#160;      logliks.resize(ll-&gt;size(),0.0);</div>
<div class="line"><a name="l00042"></a><span class="lineno">   42</span>&#160;      <span class="keywordflow">if</span>(logliks.size() == 0){</div>
<div class="line"><a name="l00043"></a><span class="lineno">   43</span>&#160;    cerr&lt;&lt;<span class="stringliteral">&quot;logliks still empty!&quot;</span>&lt;&lt;endl;</div>
<div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;      }</div>
<div class="line"><a name="l00045"></a><span class="lineno">   45</span>&#160;    }</div>
<div class="line"><a name="l00046"></a><span class="lineno">   46</span>&#160;    <span class="keywordflow">for</span>(<span class="keywordtype">int</span> u = 0; u &lt; ll-&gt;size(); u++){</div>
<div class="line"><a name="l00047"></a><span class="lineno">   47</span>&#160;      logliks[u] += (*ll)[u];</div>
<div class="line"><a name="l00048"></a><span class="lineno">   48</span>&#160;    }</div>
<div class="line"><a name="l00049"></a><span class="lineno">   49</span>&#160;  }</div>
<div class="line"><a name="l00050"></a><span class="lineno">   50</span>&#160;  <span class="keywordtype">void</span> addNumSubs(<span class="keywordtype">int</span> subs){</div>
<div class="line"><a name="l00051"></a><span class="lineno">   51</span>&#160;    numSubs += subs;</div>
<div class="line"><a name="l00052"></a><span class="lineno">   52</span>&#160;   }</div>
<div class="line"><a name="l00053"></a><span class="lineno">   53</span>&#160;};</div>
<div class="line"><a name="l00054"></a><span class="lineno">   54</span>&#160; </div>
<div class="line"><a name="l00061"></a><span class="lineno"><a class="line" href="classGeneMSA.html">   61</a></span>&#160;<span class="keyword">class </span><a class="code" href="classGeneMSA.html">GeneMSA</a> {</div>
<div class="line"><a name="l00062"></a><span class="lineno">   62</span>&#160;<span class="keyword">public</span>:</div>
<div class="line"><a name="l00063"></a><span class="lineno">   63</span>&#160;    <a class="code" href="classGeneMSA.html#ada79b45bf3f2605773220304968d5993">GeneMSA</a>(<a class="code" href="classRandSeqAccess.html">RandSeqAccess</a> *rsa, <a class="code" href="classAlignment.html">Alignment</a> *a);</div>
<div class="line"><a name="l00064"></a><span class="lineno">   64</span>&#160;    ~<a class="code" href="classGeneMSA.html">GeneMSA</a>();</div>
<div class="line"><a name="l00065"></a><span class="lineno">   65</span>&#160; </div>
<div class="line"><a name="l00066"></a><span class="lineno">   66</span>&#160;    <span class="comment">/*</span></div>
<div class="line"><a name="l00067"></a><span class="lineno">   67</span>&#160;<span class="comment">     *  get and set functions</span></div>
<div class="line"><a name="l00068"></a><span class="lineno">   68</span>&#160;<span class="comment">     */</span></div>
<div class="line"><a name="l00069"></a><span class="lineno">   69</span>&#160;    <span class="comment">// start and end position (0-based, inclusive) of the gene range for the given species</span></div>
<div class="line"><a name="l00070"></a><span class="lineno">   70</span>&#160;    <span class="comment">// start &lt;= end refer to the FORWARD strand of the sequence (different from class &#39;alignment&#39;)</span></div>
<div class="line"><a name="l00071"></a><span class="lineno">   71</span>&#160;    <span class="comment">// start = end = -1 if the species is absent from the alignment</span></div>
<div class="line"><a name="l00072"></a><span class="lineno">   72</span>&#160;    <span class="keywordtype">string</span> getSeqID(<span class="keywordtype">int</span> speciesIdx);</div>
<div class="line"><a name="l00073"></a><span class="lineno">   73</span>&#160;    Strand getStrand(<span class="keywordtype">int</span> speciesIdx);</div>
<div class="line"><a name="l00074"></a><span class="lineno">   74</span>&#160;    <span class="keywordtype">int</span> getStart(<span class="keywordtype">int</span> speciesIdx){ <span class="keywordflow">return</span> starts[speciesIdx]; }</div>
<div class="line"><a name="l00075"></a><span class="lineno">   75</span>&#160;    <span class="keywordtype">int</span> getEnd(<span class="keywordtype">int</span> speciesIdx){ <span class="keywordflow">return</span> ends[speciesIdx]; }</div>
<div class="line"><a name="l00076"></a><span class="lineno">   76</span>&#160;    list&lt;ExonCandidate*&gt;* getExonCands(<span class="keywordtype">int</span> speciesIdx){ <span class="keywordflow">return</span> exoncands.at(speciesIdx); }</div>
<div class="line"><a name="l00077"></a><span class="lineno">   77</span>&#160;    <span class="comment">//list&lt;OrthoExon&gt; getOrthoExons() { return orthoExonsList; }</span></div>
<div class="line"><a name="l00078"></a><span class="lineno">   78</span>&#160;    <a class="code" href="classAlignment.html">Alignment</a> * getAlignment() {<span class="keywordflow">return</span> alignment;}</div>
<div class="line"><a name="l00079"></a><span class="lineno">   79</span>&#160;    vector&lt;int&gt; getOffsets() {<span class="keywordflow">return</span> offsets;}</div>
<div class="line"><a name="l00080"></a><span class="lineno">   80</span>&#160; </div>
<div class="line"><a name="l00081"></a><span class="lineno">   81</span>&#160;    map&lt;string,ExonCandidate*&gt;* getECHash(list&lt;ExonCandidate*&gt; *ec); <span class="comment">// adds the keys to the map function</span></div>
<div class="line"><a name="l00082"></a><span class="lineno">   82</span>&#160; </div>
<div class="line"><a name="l00083"></a><span class="lineno">   83</span>&#160;    <span class="comment">/*</span></div>
<div class="line"><a name="l00084"></a><span class="lineno">   84</span>&#160;<span class="comment">     * assmotifqthresh, assqthresh, dssqthresh are between 0 and 1 and thresholds for the inclusion of</span></div>
<div class="line"><a name="l00085"></a><span class="lineno">   85</span>&#160;<span class="comment">     * acceptor/donor splice sites based on the pattern probability</span></div>
<div class="line"><a name="l00086"></a><span class="lineno">   86</span>&#160;<span class="comment">     * assqthresh=0.05 means that only acceptor ss are considered</span></div>
<div class="line"><a name="l00087"></a><span class="lineno">   87</span>&#160;<span class="comment">     * that have a pattern, such that 5% of true splice site patterns have lower probability.</span></div>
<div class="line"><a name="l00088"></a><span class="lineno">   88</span>&#160;<span class="comment">     * The default threshold of 0 means that all splice site patterns are considered.</span></div>
<div class="line"><a name="l00089"></a><span class="lineno">   89</span>&#160;<span class="comment">     */</span></div>
<div class="line"><a name="l00090"></a><span class="lineno">   90</span>&#160;    <span class="keywordtype">void</span> createExonCands(<span class="keywordtype">int</span> s, <span class="keyword">const</span> <span class="keywordtype">char</span> *dna, map&lt;int_fast64_t, ExonCandidate*&gt; &amp;ecs, map&lt;int_fast64_t, ExonCandidate*&gt; &amp;addECs);</div>
<div class="line"><a name="l00091"></a><span class="lineno">   91</span>&#160;    <span class="keywordtype">void</span> setExonCands(vector&lt;map&lt;int_fast64_t, ExonCandidate*&gt; &gt; &amp;ecs);</div>
<div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160; </div>
<div class="line"><a name="l00103"></a><span class="lineno">  103</span>&#160;    <span class="keywordtype">void</span> <a class="code" href="classGeneMSA.html#aad55bc5eb24509be6b5aa056e8db5ef1">createOrthoExons</a>(list&lt;OrthoExon&gt; &amp;orthoExonsList, map&lt;int_fast64_t, list&lt;pair&lt;int,ExonCandidate*&gt; &gt; &gt; &amp;alignedECs, <a class="code" href="classEvo.html">Evo</a> *evo, <span class="keywordtype">float</span> consThres = 0.0, <span class="keywordtype">int</span> minAvLen = 0);</div>
<div class="line"><a name="l00104"></a><span class="lineno">  104</span>&#160;    <span class="keywordtype">void</span> printStats(); <span class="comment">// to stdout</span></div>
<div class="line"><a name="l00105"></a><span class="lineno">  105</span>&#160;    <span class="keywordtype">void</span> printGeneRanges();</div>
<div class="line"><a name="l00106"></a><span class="lineno">  106</span>&#160;    <span class="keywordtype">void</span> printExonCands();</div>
<div class="line"><a name="l00107"></a><span class="lineno">  107</span>&#160;    <span class="keywordtype">void</span> printOrthoExons(list&lt;OrthoExon&gt; &amp;orthoExonsList);</div>
<div class="line"><a name="l00108"></a><span class="lineno">  108</span>&#160;    <span class="keywordtype">void</span> computeOmegas(list&lt;OrthoExon&gt; &amp;orthoExonsList, vector&lt;AnnoSequence*&gt; <span class="keyword">const</span> &amp;seqRanges, <a class="code" href="classPhyloTree.html">PhyloTree</a> *ctree);</div>
<div class="line"><a name="l00109"></a><span class="lineno">  109</span>&#160;    <span class="keywordtype">void</span> computeOmegasEff(list&lt;OrthoExon&gt; &amp;orthoExonsList, vector&lt;AnnoSequence*&gt; <span class="keyword">const</span> &amp;seqRanges, <a class="code" href="classPhyloTree.html">PhyloTree</a> *ctree, ofstream *codonAli);</div>
<div class="line"><a name="l00110"></a><span class="lineno">  110</span>&#160;    vector&lt;string&gt; pruneToBV(vector&lt;string&gt; *cs, bit_vector bv); <span class="comment">// prune codon strings to bit_vector</span></div>
<div class="line"><a name="l00111"></a><span class="lineno">  111</span>&#160;    vector&lt;int&gt; pruneToBV(vector&lt;int&gt; *rfc, bit_vector bv); <span class="comment">// prune RFC to bit_vector</span></div>
<div class="line"><a name="l00112"></a><span class="lineno">  112</span>&#160;    <span class="keywordtype">double</span> omegaForCodonTuple(vector&lt;double&gt; *loglik);</div>
<div class="line"><a name="l00113"></a><span class="lineno">  113</span>&#160;    <span class="keywordtype">void</span> printOmegaForCodon(<span class="keywordtype">string</span> outdir);</div>
<div class="line"><a name="l00114"></a><span class="lineno">  114</span>&#160;    <span class="keywordtype">void</span> printCumOmega();</div>
<div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;    <span class="keywordtype">void</span> comparativeSignalScoring(list&lt;OrthoExon&gt; &amp;orthoExonsList);</div>
<div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;    <span class="comment">// Charlotte Janas playground</span></div>
<div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;    <a class="code" href="classLocusTree.html">LocusTree</a> *constructTree(); <span class="comment">// creates, stores are returns the locus tree for the sequence tuple</span></div>
<div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160; </div>
<div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;    <span class="comment">// calculate a columnwise conservation score and output it (for each species) in wiggle format</span></div>
<div class="line"><a name="l00120"></a><span class="lineno">  120</span>&#160;    <span class="keywordtype">void</span> calcConsScore(list&lt;OrthoExon&gt; &amp;orthoExonsList, vector&lt;AnnoSequence*&gt; <span class="keyword">const</span> &amp;seqRanges, <span class="keywordtype">string</span> outdir);</div>
<div class="line"><a name="l00121"></a><span class="lineno">  121</span>&#160;    <span class="keywordtype">double</span> calcColumnScore(<span class="keywordtype">int</span> a, <span class="keywordtype">int</span> c, <span class="keywordtype">int</span> t, <span class="keywordtype">int</span> g); <span class="comment">// input: number of a,c,t and g&#39;s in one alignment column </span></div>
<div class="line"><a name="l00122"></a><span class="lineno">  122</span>&#160;    <span class="keywordtype">void</span> consToWig(vector&lt;double&gt; &amp;consScore, <span class="keywordtype">string</span> outdir);</div>
<div class="line"><a name="l00123"></a><span class="lineno">  123</span>&#160; </div>
<div class="line"><a name="l00124"></a><span class="lineno">  124</span>&#160;    <span class="comment">// static functions</span></div>
<div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">void</span> setTree(<a class="code" href="classPhyloTree.html">PhyloTree</a> *t){tree = t;}</div>
<div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">void</span> setCodonEvo(<a class="code" href="classCodonEvo.html">CodonEvo</a> *c){ codonevo = c; }</div>
<div class="line"><a name="l00127"></a><span class="lineno">  127</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">int</span> numSpecies(){ <span class="keywordflow">return</span> tree-&gt;numSpecies(); }</div>
<div class="line"><a name="l00128"></a><span class="lineno">  128</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">void</span> openOutputFiles(<span class="keywordtype">string</span> outdir);</div>
<div class="line"><a name="l00129"></a><span class="lineno">  129</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">void</span> closeOutputFiles();</div>
<div class="line"><a name="l00130"></a><span class="lineno">  130</span>&#160; </div>
<div class="line"><a name="l00131"></a><span class="lineno">  131</span>&#160;    <span class="comment">// static data members</span></div>
<div class="line"><a name="l00132"></a><span class="lineno">  132</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">int</span> padding; <span class="comment">// add this many bases to the region before and after the aligned region</span></div>
<div class="line"><a name="l00133"></a><span class="lineno">  133</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">int</span> orthoExonID; <span class="comment">// ID for exons of different species which are orthologous to each other</span></div>
<div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;    <span class="keyword">static</span> <span class="keywordtype">int</span> geneRangeID; <span class="comment">// stores an ID for the possible gene ranges of the different species which belong together</span></div>
<div class="line"><a name="l00135"></a><span class="lineno">  135</span>&#160;    <span class="keyword">static</span> vector&lt;int&gt; exonCandID; <span class="comment">// IDs for exon candidates of different species</span></div>
<div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;    <span class="keyword">static</span> unordered_map&lt; bit_vector, PhyloTree*, boost::hash&lt;bit_vector&gt;&gt; topologies;</div>
<div class="line"><a name="l00137"></a><span class="lineno">  137</span>&#160;    <span class="comment">// pointers to the output files</span></div>
<div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;    <span class="keyword">static</span> vector&lt; ofstream* &gt; exonCands_outfiles, orthoExons_outfiles, geneRanges_outfiles_bed, geneRanges_outfiles_gff, omega_outfiles; </div>
<div class="line"><a name="l00139"></a><span class="lineno">  139</span>&#160;    <span class="comment">// map that stored all codon combinations on which fitch and pruning algorithm already have been run (for calculation of omega and number of substitutions)</span></div>
<div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;    <span class="keyword">static</span> map&lt;vector&lt;string&gt;, pair&lt;vector&lt;double&gt;, <span class="keywordtype">int</span>&gt; &gt; computedCumValues;</div>
<div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160; </div>
<div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;    <span class="keywordtype">void</span> printSingleOrthoExon(<a class="code" href="classOrthoExon.html">OrthoExon</a> &amp;oe, <span class="keywordtype">bool</span> files = <span class="keyword">true</span>);</div>
<div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;    <span class="keywordtype">void</span> collect_features(<span class="keywordtype">int</span> species, list&lt;OrthoExon&gt; *hects, <a class="code" href="classSpeciesGraph.html">SpeciesGraph</a> *speciesgraph);</div>
<div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160; </div>
<div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;    <span class="keywordtype">void</span> <a class="code" href="classGeneMSA.html#ae023b0e02bb2a77a2b87db9126384f91">getAllOEMsas</a>(<span class="keywordtype">int</span> species, list&lt;OrthoExon&gt; *hects, unordered_map&lt;string,int&gt; *ref_class,</div>
<div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;                         vector&lt;AnnoSequence*&gt; <span class="keyword">const</span> &amp;seqRanges);</div>
<div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160; </div>
<div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;    <a class="code" href="classStringAlignment.html">StringAlignment</a> <a class="code" href="classGeneMSA.html#abd42e0ecb942aceb3d0b7d7f4c104dce">getMsa</a>(<a class="code" href="classOrthoExon.html">OrthoExon</a> <span class="keyword">const</span> &amp;oe, vector&lt;AnnoSequence*&gt; <span class="keyword">const</span> &amp;seqRanges, <span class="keywordtype">size_t</span> flanking = 0);</div>
<div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;<span class="keyword">private</span>:</div>
<div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;  vector&lt;string&gt; getCodonAlignment(<a class="code" href="classOrthoExon.html">OrthoExon</a> <span class="keyword">const</span> &amp;oe, vector&lt;AnnoSequence*&gt; <span class="keyword">const</span> &amp;seqRanges, <span class="keyword">const</span> vector&lt;vector&lt;fragment&gt;::const_iterator &gt; &amp;froms, map&lt;<span class="keywordtype">unsigned</span>, vector&lt;int&gt; &gt; *alignedCodons = NULL, <span class="keywordtype">bool</span> generateString=<span class="keyword">true</span>, vector&lt;vector&lt;int&gt; &gt; *posStoredCodons = NULL, ofstream *codonAli = NULL);</div>
<div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;    <span class="keywordtype">void</span> cutIncompleteCodons(<a class="code" href="classOrthoExon.html">OrthoExon</a> &amp;oe);</div>
<div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;    <a class="code" href="structcumValues.html">cumValues</a>* findCumValues(bit_vector bv, vector&lt;int&gt; rfc);</div>
<div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;    <span class="keyword">static</span> <a class="code" href="classPhyloTree.html">PhyloTree</a> *tree;</div>
<div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;    <a class="code" href="classLocusTree.html">LocusTree</a> *ltree;</div>
<div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;    <span class="keyword">static</span> <a class="code" href="classCodonEvo.html">CodonEvo</a> *codonevo;</div>
<div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;    vector&lt;int&gt; starts, ends; <span class="comment">// gene ranges for each species</span></div>
<div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;    vector&lt;int&gt; offsets; <span class="comment">// this many bases are upstream from the region</span></div>
<div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;    <a class="code" href="classRandSeqAccess.html">RandSeqAccess</a> *rsa;</div>
<div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;    <a class="code" href="classAlignment.html">Alignment</a>* alignment;            <span class="comment">// alignment of regions which possibly belong to a gene</span></div>
<div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;    vector&lt; list&lt;ExonCandidate*&gt;* &gt; exoncands;  <span class="comment">// exon candidates found in the different species in a gene segment</span></div>
<div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;    <span class="comment">//list&lt;OrthoExon&gt; orthoExonsList;       // Steffi: due to multiple copying, I remove this as attribute of any class. Instead it can be passed from compgenepred.cc by reference, whenever necessary.</span></div>
<div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;    unordered_map&lt;bit_vector, vector&lt;pair&lt;vector&lt;int&gt;, <a class="code" href="structcumValues.html">cumValues</a>&gt; &gt;, boost::hash&lt;bit_vector&gt; &gt; cumOmega; <span class="comment">// stores cumulative omega values for every reading frame combination and every bitvector that exist </span></div>
<div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;    map&lt;bit_vector, map&lt;vector&lt;int&gt;, vector&lt;double&gt; &gt; &gt; codonOmega;</div>
<div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160; </div>
<div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;<span class="preprocessor">    #ifdef TESTING</span></div>
<div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;    <span class="comment">/*</span></div>
<div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;<span class="comment">    *   added by Giovanna Migliorelli 14.05.2020 </span></div>
<div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;<span class="comment">    *   code responsible for serialization</span></div>
<div class="line"><a name="l00187"></a><span class="lineno">  187</span>&#160;<span class="comment">    *   aknowledgement : https://www.boost.org/doc/libs/1_70_0/libs/serialization/doc/tutorial.html</span></div>
<div class="line"><a name="l00188"></a><span class="lineno">  188</span>&#160;<span class="comment">    */</span></div>
<div class="line"><a name="l00189"></a><span class="lineno">  189</span>&#160; </div>
<div class="line"><a name="l00190"></a><span class="lineno">  190</span>&#160;    <span class="keyword">friend</span> <span class="keyword">class </span>boost::serialization::access;</div>
<div class="line"><a name="l00191"></a><span class="lineno">  191</span>&#160;    <span class="keyword">template</span>&lt;<span class="keyword">class</span> Archive&gt;</div>
<div class="line"><a name="l00192"></a><span class="lineno">  192</span>&#160;    <span class="keywordtype">void</span> serialize(Archive &amp; ar, <span class="keyword">const</span> <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> version)</div>
<div class="line"><a name="l00193"></a><span class="lineno">  193</span>&#160;    {</div>
<div class="line"><a name="l00194"></a><span class="lineno">  194</span>&#160;        ar &amp; starts;</div>
<div class="line"><a name="l00195"></a><span class="lineno">  195</span>&#160;        ar &amp; alignment;</div>
<div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160; </div>
<div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;    }</div>
<div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160; </div>
<div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;    <span class="comment">/*  added by Giovanna Migliorelli 14.05.2020</span></div>
<div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;<span class="comment">    *   default consttruct to allow for serialization</span></div>
<div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;<span class="comment">    */</span></div>
<div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160; </div>
<div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;    <a class="code" href="classGeneMSA.html#ada79b45bf3f2605773220304968d5993">GeneMSA</a>(){}; </div>
<div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160; </div>
<div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;<span class="preprocessor">    #endif</span></div>
<div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;};</div>
<div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160; </div>
<div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160; </div>
<div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160; </div>
<div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;<span class="preprocessor">#endif  </span><span class="comment">// _GENEMSA</span></div>
<div class="ttc" id="aclassAlignment_html"><div class="ttname"><a href="classAlignment.html">Alignment</a></div><div class="ttdoc">global multiple sequence alignment with efficiently stored long gaps.</div><div class="ttdef"><b>Definition:</b> alignment.hh:160</div></div>
<div class="ttc" id="aclassCodonEvo_html"><div class="ttname"><a href="classCodonEvo.html">CodonEvo</a></div><div class="ttdoc">class CodonEvo</div><div class="ttdef"><b>Definition:</b> codonevo.hh:30</div></div>
<div class="ttc" id="aclassEvo_html"><div class="ttname"><a href="classEvo.html">Evo</a></div><div class="ttdoc">abstract base class for codon and exon evolution</div><div class="ttdef"><b>Definition:</b> contTimeMC.hh:31</div></div>
<div class="ttc" id="aclassGeneMSA_html"><div class="ttname"><a href="classGeneMSA.html">GeneMSA</a></div><div class="ttdoc">multiple sequence alignment of genomes for comparative gene prediction</div><div class="ttdef"><b>Definition:</b> geneMSA.hh:61</div></div>
<div class="ttc" id="aclassGeneMSA_html_aad55bc5eb24509be6b5aa056e8db5ef1"><div class="ttname"><a href="classGeneMSA.html#aad55bc5eb24509be6b5aa056e8db5ef1">GeneMSA::createOrthoExons</a></div><div class="ttdeci">void createOrthoExons(list&lt; OrthoExon &gt; &amp;orthoExonsList, map&lt; int_fast64_t, list&lt; pair&lt; int, ExonCandidate * &gt; &gt; &gt; &amp;alignedECs, Evo *evo, float consThres=0.0, int minAvLen=0)</div><div class="ttdef"><b>Definition:</b> geneMSA.cc:222</div></div>
<div class="ttc" id="aclassGeneMSA_html_abd42e0ecb942aceb3d0b7d7f4c104dce"><div class="ttname"><a href="classGeneMSA.html#abd42e0ecb942aceb3d0b7d7f4c104dce">GeneMSA::getMsa</a></div><div class="ttdeci">StringAlignment getMsa(OrthoExon const &amp;oe, vector&lt; AnnoSequence * &gt; const &amp;seqRanges, size_t flanking=0)</div><div class="ttdef"><b>Definition:</b> geneMSA.cc:767</div></div>
<div class="ttc" id="aclassGeneMSA_html_ada79b45bf3f2605773220304968d5993"><div class="ttname"><a href="classGeneMSA.html#ada79b45bf3f2605773220304968d5993">GeneMSA::GeneMSA</a></div><div class="ttdeci">GeneMSA(RandSeqAccess *rsa, Alignment *a)</div><div class="ttdef"><b>Definition:</b> geneMSA.cc:43</div></div>
<div class="ttc" id="aclassGeneMSA_html_ae023b0e02bb2a77a2b87db9126384f91"><div class="ttname"><a href="classGeneMSA.html#ae023b0e02bb2a77a2b87db9126384f91">GeneMSA::getAllOEMsas</a></div><div class="ttdeci">void getAllOEMsas(int species, list&lt; OrthoExon &gt; *hects, unordered_map&lt; string, int &gt; *ref_class, vector&lt; AnnoSequence * &gt; const &amp;seqRanges)</div><div class="ttdef"><b>Definition:</b> geneMSA.cc:731</div></div>
<div class="ttc" id="aclassLocusTree_html"><div class="ttname"><a href="classLocusTree.html">LocusTree</a></div><div class="ttdef"><b>Definition:</b> phylotree.hh:173</div></div>
<div class="ttc" id="aclassOrthoExon_html"><div class="ttname"><a href="classOrthoExon.html">OrthoExon</a></div><div class="ttdoc">maintains orthologous exons for comparative gene prediction</div><div class="ttdef"><b>Definition:</b> orthoexon.hh:35</div></div>
<div class="ttc" id="aclassOrthoGraph_html"><div class="ttname"><a href="classOrthoGraph.html">OrthoGraph</a></div><div class="ttdoc">orthologous graphs for comparative gene prediction</div><div class="ttdef"><b>Definition:</b> orthograph.hh:34</div></div>
<div class="ttc" id="aclassPhyloTree_html"><div class="ttname"><a href="classPhyloTree.html">PhyloTree</a></div><div class="ttdef"><b>Definition:</b> phylotree.hh:106</div></div>
<div class="ttc" id="aclassRandSeqAccess_html"><div class="ttname"><a href="classRandSeqAccess.html">RandSeqAccess</a></div><div class="ttdoc">abstract class for quick access to an arbitrary sequence segment in genomes needed for comparative ge...</div><div class="ttdef"><b>Definition:</b> randseqaccess.hh:62</div></div>
<div class="ttc" id="aclassSpeciesGraph_html"><div class="ttname"><a href="classSpeciesGraph.html">SpeciesGraph</a></div><div class="ttdoc">builds a directed acyclic graph from a set of sampled genes.</div><div class="ttdef"><b>Definition:</b> speciesgraph.hh:32</div></div>
<div class="ttc" id="aclassStringAlignment_html"><div class="ttname"><a href="classStringAlignment.html">StringAlignment</a></div><div class="ttdoc">global multiple sequence alignment in (standard) string representation</div><div class="ttdef"><b>Definition:</b> alignment.hh:351</div></div>
<div class="ttc" id="astructcumValues_html"><div class="ttname"><a href="structcumValues.html">cumValues</a></div><div class="ttdef"><b>Definition:</b> geneMSA.hh:34</div></div>
</div><!-- fragment --></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.0
</small></address>
</body>
</html>
